src/events/attributes.rs

```diff
@@ -331,120 +327,26 @@ impl<'a> Attributes<'a> {
-            };
-        }
-
-        if len <= self.position {
-            return None;
-        }
-
-        let mut bytes = self.bytes.iter().enumerate().skip(self.position);
-
-        // key starts after the whitespace
-        let start_key = match bytes
-            .by_ref()
-            .skip_while(|&(_, &b)| !is_whitespace(b))
-            .find(|&(_, &b)| !is_whitespace(b))
-        {
-            Some((i, _)) => i,
-            None => return attr!(self.position..len),
-        };
-
-        // key ends with either whitespace or =
-        let end_key = match bytes
-            .by_ref()
-            .find(|&(_, &b)| b == b'=' || is_whitespace(b))
-        {
-            Some((i, &b'=')) => i,
-            Some((i, _)) => {
-                // consume until `=` or return if html
-                match bytes.by_ref().find(|&(_, &b)| !is_whitespace(b)) {
-                    Some((_, &b'=')) => i,
-                    Some((j, _)) if self.html => {
-                        self.position = j - 1;
-                        return attr!(start_key..i, 0..0);
-                    }
-                    Some((j, _)) => err!(AttrError::ExpectedEq(j)),
-                    None if self.html => {
-                        self.position = len;
-                        return attr!(start_key..len, 0..0);
-                    }
-                    None => err!(AttrError::ExpectedEq(len)),
-                }
-            }
-            None => return attr!(start_key..len),
-        };
-
-        if self.with_checks {
-            if let Some(start) = self
-                .consumed
-                .iter()
-                .filter(|r| r.len() == end_key - start_key)
-                .find(|r| self.bytes[(*r).clone()] == self.bytes[start_key..end_key])
-                .map(|ref r| r.start)
-            {
-                err!(AttrError::Duplicated(start_key, start));
-            }
-            self.consumed.push(start_key..end_key);
-        }
-
-        // value has quote if not html
-        match bytes.by_ref().find(|&(_, &b)| !is_whitespace(b)) {
-            Some((i, quote @ &b'\'')) | Some((i, quote @ &b'"')) => {
-                match bytes.by_ref().find(|&(_, &b)| b == *quote) {
-                    Some((j, _)) => {
-                        self.position = j + 1;
-                        return attr!(start_key..end_key, i + 1..j);
-                    }
-                    None => err!(AttrError::UnquotedValue(i)),
-                }
-            }
-            Some((i, _)) if self.html => {
-                let j = bytes
-                    .by_ref()
-                    .find(|&(_, &b)| is_whitespace(b))
-                    .map_or(len, |(j, _)| j);
-                self.position = j;
-                return attr!(start_key..end_key, i..j);
-            }
-            Some((i, _)) => err!(AttrError::UnquotedValue(i)),
-            None => return attr!(start_key..end_key),
+        match self.state.next(self.bytes) {
```

Comment:
This is required for lifetime correctness. Owned iterator will store a `Cow<'input, [u8]>`, borrowing iterator stores a `&'input [u8]`. If you will try to move `slice` argument of `next` into `IterState` and store in a `Cow<'input, [u8]>`, even when borrowing iterator always will pass a `Cow::Borrowed`, borrow checker will not allow you to return slices with the `'input` lifetime, because `Cow` potentially can be owned.

You also cannot store a `&'input [u8]` in the `IterState`, because owned iterator is supposed to be
```rust
struct OwnedAttributes<'input> {
  slice: Cow<'input, [u8]>,
  state: IterState,
}
```
and you cannot store in that struct reference to a field of this struct (`state` would be referenced to a `slice`)